File: src/options.cpp
//
// Created by Dennis Sitelew on 22.12.22.
//

#include <ocs/options.h>
#include <ocs/video.h>

#include <boost/filesystem.hpp>
#include <lyra/lyra.hpp>

#include <thread>

using namespace ocs;

namespace {

std::string get_default_tess_data_path(const char *app_path) {
   boost::filesystem::path path{app_path};
   path.remove_filename();
   path /= "tessdata";
   return path.string();
}

} // namespace

std::optional<options> options::parse(int argc, const char **argv) {
   options result;
   result.ocr_threads = std::thread::hardware_concurrency();
   result.language = "eng+rus+deu";
   result.frame_filter = static_cast<std::uint16_t>(video::frame_filter::I_and_P);
   result.tess_data_path = get_default_tess_data_path(argv[0]);

   bool show_help{false};

   auto cli =
       lyra::opt(result.ocr_threads, "num_threads")["-p"]["--num-threads"]("Number of threads to use for OCR") |
       lyra::opt(result.language, "language")["-l"]["--language"]("OCR language, e.g. 'eng+rus+deu', or just 'eng'") |
       lyra::opt(result.tess_data_path, "tess_data_path")["-t"]["--tess-data-path"](
           "Path to the Tesseract data directory. You can download a copy here: "
           "https://github.com/tesseract-ocr/tessdata/releases/") |
       lyra::opt(result.frame_filter, "frame_filter")["-f"]["--frame-filter"](
           "Video frame filter. 1 for I-frames, 2 for P-frames, and 4 for B-frames, can be a combination of multiple "
           "frame types. 3 is default (I+P frames)") |
       lyra::opt(result.video_file, "video_file")["-i"]["--video-file"]("Video file to process").required() |
       lyra::opt(result.database_file, "database_file")["-o"]["--database-file"]("Resulting OCR database").required() |
       lyra::opt([&](bool) { result.save_bitmaps = true; })["-b"]["--save-bitmaps"](
           "Save video bitmaps in the out/ subdirectory") |
       lyra::help(show_help);

   auto parse_result = cli.parse({argc, argv});
   if (!parse_result) {
      std::cerr << "Error in command line: " << parse_result.message() << std::endl;
      std::cerr << cli << std::endl;
      return std::nullopt;
   }

   if (!boost::filesystem::exists(result.video_file)) {
      std::cerr << "Video file does not exist: " << result.video_file << std::endl;
      return std::nullopt;
   }

   if (!boost::filesystem::exists(result.tess_data_path)) {
      std::cerr << "Tesseract data directory does not exist: " << result.tess_data_path << std::endl;
      return std::nullopt;
   }

   if (show_help) {
      std::cerr << cli << std::endl;
      return {};
   }

   return result;
}

File: tools/ocs-watcher/ocsw/options.py
import argparse
import os


class Options:
    def __init__(self, args):
        self.ocr_suite = args.ocr_suite
        self.ocr_lang = args.ocr_lang
        self.tess_data_path = args.tess_data_path
        self.num_ocr_threads = args.num_ocr_threads

        self.video_dir = args.video_dir
        self.check_frequency = args.checking_frequency
        self.file_extension = args.file_extension

    @staticmethod
    def get_options():
        num_threads = max(os.cpu_count() - 2, 1)

        parser = argparse.ArgumentParser()
        parser.add_argument('-o', '--ocr-suite', required=True, help='Path to the ocr-suite executable')
        parser.add_argument('-i', '--video-dir', required=True, help='Video files directory')
        parser.add_argument('-t', '--tess-data-path', required=True,
                            help='Path to the Tesseract data directory. You can download a copy here: '
                                 'https://github.com/tesseract-ocr/tessdata/releases/')
        parser.add_argument('-p', '--num-ocr-threads', default=num_threads, help='Number of OCR threads')
        parser.add_argument('-l', '--ocr-lang', default='eng+rus+deu',
                            help="OCR language, e.g. 'eng+rus+deu', or just 'eng'")
        parser.add_argument('-e', '--file-extension', type=str, default='mkv', help='Video files extension')
        parser.add_argument('-f', '--checking-frequency', type=int, default=10,
                            help='Video recognition frequency in minutes (can be used to recognize a vidio, '
                                 'which is still being recorded).')

        args = parser.parse_args()
        return Options(args)

    def __str__(self):
        params = {
            'ocr_suite': self.ocr_suite,
            'video_dir': self.video_dir,
            'tess_data_path': self.tess_data_path,
            'num_ocr_threads': self.num_ocr_threads,
            'ocr_lang': self.ocr_lang,
            'check_frequency': self.check_frequency,
            'file_extension': self.file_extension,
        }
        return str(params)

    def __repr__(self):
        return self.__str__()
